File: src/cdr.ts
// Call history (CDR) for the logged-in operator, fetched via the BFF (which
// scopes the query to the caller's own extension).
import { bffUrl } from "./config";

export interface CdrRow {
  id: string;
  direction: string;
  caller_id_number: string;
  caller_id_name: string;
  destination_number: string;
  hangup_cause: string;
  start_epoch: number;
  answer_epoch: number;
  billsec: number;
}

export async function fetchCdr(token: string, limit = 50): Promise<{ cdrs: CdrRow[]; total: number }> {
  const r = await fetch(`${bffUrl}/api/cdr?limit=${limit}`, {
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!r.ok) throw new Error(`call history failed (${r.status})`);
  return r.json();
}

File: bff/cdr.go
package main

import (
	"encoding/json"
	"errors"
	"net/http"
	"net/url"
	"strconv"
)

// handleCDR returns the logged-in operator's own call history. The operator's
// extension is injected as the `number` filter, so an agent only ever sees
// their own calls — the browser cannot query arbitrary numbers.
func (s *server) handleCDR(w http.ResponseWriter, r *http.Request) {
	id := identityFrom(r.Context())

	op, err := s.cp.operator(r.Context(), id.Subject)
	if errors.Is(err, errNotFound) {
		writeErr(w, http.StatusForbidden, "no operator binding for this identity")
		return
	}
	if err != nil {
		writeErr(w, http.StatusBadGateway, "operator lookup failed: "+err.Error())
		return
	}

	limit := clampInt(r.URL.Query().Get("limit"), 50, 1, 200)
	offset := clampInt(r.URL.Query().Get("offset"), 0, 0, 1<<30)

	q := url.Values{}
	q.Set("number", op.Number)
	q.Set("limit", strconv.Itoa(limit))
	q.Set("offset", strconv.Itoa(offset))

	body, total, err := s.cp.getWithTotal(r.Context(), "/api/v1/cdr?"+q.Encode())
	if err != nil {
		writeErr(w, http.StatusBadGateway, "cdr lookup failed: "+err.Error())
		return
	}

	// Pass the CDR array through verbatim; embed total in the body (the browser
	// can't read X-Total-Count without a CORS expose-headers allowance).
	writeJSON(w, http.StatusOK, map[string]any{"cdrs": json.RawMessage(body), "total": total})
}
